### DISCUSSION TITLE
Using sockets/TCP in add-ons

### DISCUSSION THREAD

**Discussion by CasperClous** [lieser/dkim_verifier#244] [Q&A]
Hello, I'm trying to make a new add-on for TB, and I have some problems, for some reason did you use socket TCP in this add-on? I'm trying to make a tcp connection, but I cant find the way, if you could help me I will be very grateful.

**Comment by tadpole2077** [comment]
Hi

Correct the DNS public server query's all use TCP connections, as initially the inherited library was written for TCP only (so no UDP usage)
This DNS library was originally developed by Joshua Tauberer for "Sender Verification Extension" version 0.9.0.6,  then later Philippe Lieser extended-recoded for DKIM plugin usage.

Try reviewing library \experiments\JSDNS.js.

Or walk though of code use by breakpoint in TB Developer Toolbox >> Main Thread >jar:// >users/xxxxx > dkim_verifier .. > JSDNS.js
* configureDNS() 
* txt(name) ...  dnsCallback  Ln 72 breakpoint , to trace the actual query to the selected DNS to pull the TXT record containing the DKIM public key and selector.

**Comment by lieser** [comment]
Is it really raw TCP that you need, instead of HTTP? For HTTP there exist the `webRequest` API https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest.

As raw TCP is not something that is currently provided by the MailExtension API, you have to use a WebExtension Experiments for it.
See also:
- https://developer.thunderbird.net/add-ons/mailextensions/experiments
- https://webextension-api.thunderbird.net/en/latest/how-to/experiments.html#

As @tadpole2077 already wrote, the TCP code is in `experiments/JSDNS.jsm`, more specific the `DNS_readAllFromSocket`  function inside that file.
https://github.com/lieser/dkim_verifier/blob/87d68740a0a84c54c6ca513efe94ed54a240a3ff/experiments/JSDNS.jsm.js#L955

Note that Dirk Steinmetz created a more general experiment API for TCP, that may be more useful for you: https://git.generalsync.com/oss/thunderbird-experiments.git/blob/HEAD:/experiments/tcp/child.js ([GitHub URL](https://github.com/rsjtdrjgfuzkfg/thunderbird-experiments/blob/master/experiments/tcp/child.js))


I would strongly suggest you write your question too the *​Add-On Developer Topicbox* mailing list (see https://developer.thunderbird.net/add-ons/community), as that place is more suited for general questions about add-on development.
I would also include your actual use case, i.e. that are your actually trying to do (for which you need TCP)?
Also helpful would be some example code of what you already tried.

**Comment by lieser** [comment]
Closing as more being a question about add-on development that the add-on.
